### DISCUSSION TITLE
How to get an awaited validation schema before passing to an action ?

### DISCUSSION THREAD

**Discussion by jeltehomminga** [next-safe-action/next-safe-action#111] [Q&A] [answered]
First of all thanks for this great library!

I am using translated error messages in my zod Schema. To make my schema shareable between client and server I pass the translation function to a getValidationSchema function. To get my translations on the server side I first have to await my getTranslations function. I tried making a wrapper to do this, but got stuck and I am confused how to use the action after I got my shema in my async function

If someone know how I could do this please let me know. This is how my server action looks like

```typescript
type EmailFormData = {
  newEmail: string
  confirmEmail: string
}

export type FormState = ZodIssue[] | EmailFormData

export const onSubmitEmailAction = async (prevData: FormState, actionData: EmailFormData) => {
  const { t } = await getTranslations('nl', 'myProfile')
  const schema = getValidationSchema(t)

  // how to use the action from the safe-action library here ?
}
```

**Comment by ali-alharthi** [comment]
I am facing the same issue. We use the same approach to translate zod validation errors rather than `z.setErrorMap(...)` which is more consistence for us and allows more customizable error translation IMO.
@TheEdoRan Any suggestions/plans to solve/support this?

Edit:
I believe if the schema supports async functions this would be solved (maybe?)
```js
export const signIn = guestActionClient.
  schema(
    async () => {
      const t= await getTranslations('validation')
      return signInSchema(t)
    })
  .action(
    ...
  )
```

**Reply by TheEdoRan** [reply]
Hi, where does `getTranslations()` come from? I'm not familiar with Zod + i18n, if you could please put up an example repo of your i18n setup without next-safe-action, that would surely help me understand this a little better, thanks!

**Reply by ali-alharthi** [reply]
Thanks for the quick response. So the `getTranslations()` comes from any i18n provider for our case we use [next-intl](https://next-intl-docs.vercel.app/docs/environments/server-client-components#async-components). I will try to create a working repo today to show-case how it works. Again thanks for the reply and this awesome package!

**Reply by ali-alharthi** [reply]
@TheEdoRan here is an [example](https://github.com/ali-alharthi/i18n-example) hope it helps.

**Reply by TheEdoRan** [reply]
Thank you @ali-alharthi! So the problem appears to be that i18n JSON files are dynamically imported in `getRequestConfig()` and so `getTranslations()` from `next-intl/server` is async and has to be awaited, right?

**Reply by ali-alharthi** [reply]
@TheEdoRan correct, it must be awaited.

**Reply by TheEdoRan** [reply]
Will look into it soon, probably after v7 release. Thanks!

**Reply by ali-alharthi** [reply]
Understandable, thanks again for your time👍🏼

**Reply by TheEdoRan** [reply]
Quick update: I'm working on this feature right now, expect it to be implemented very soon in the library.

**Comment by TheEdoRan** [comment, marked as answer]
Implemented this feature in [v7.0.0-next.51](https://github.com/TheEdoRan/next-safe-action/releases/tag/v7.0.0-next.51), and added a [i18n recipe](https://next.next-safe-action.dev/docs/recipes/i18n) on the website too, to explain how this works in detail.
This will hopefully solve this issue and enable support for i18n solutions, but if you still experience problems feel free to leave a comment down here and I'll reopen this discussion, thanks!

**Reply by amannn** [reply]
A related question about localizing error messages came up here: https://github.com/amannn/next-intl/discussions/1134. I'm not a zod expert, but from a first look, it seems like passing a contextual `errorMap` would work great here ([example](https://github.com/amannn/next-intl/discussions/1134#discussioncomment-9807622)).

Does `next-safe-action` support this option from `zod`?
